include/rapidjson/allocators.h

```diff
@@ -449,29 +459,50 @@ class StdAllocator :
     ~StdAllocator() RAPIDJSON_NOEXCEPT
     { }
 
-    typedef typename allocator_type::value_type       value_type;
-    typedef typename allocator_type::pointer          pointer;
-    typedef typename allocator_type::const_pointer    const_pointer;
-    typedef typename allocator_type::reference        reference;
-    typedef typename allocator_type::const_reference  const_reference;
-    typedef typename allocator_type::size_type        size_type;
-    typedef typename allocator_type::difference_type  difference_type;
-
     template<typename U>
     struct rebind {
         typedef StdAllocator<U, BaseAllocator> other;
     };
 
+    typedef typename traits_type::size_type         size_type;
+    typedef typename traits_type::difference_type   difference_type;
+    typedef typename traits_type::value_type        value_type;
+    typedef typename traits_type::pointer           pointer;
+    typedef typename traits_type::const_pointer     const_pointer;
+
 #if RAPIDJSON_HAS_CXX11
-    using allocator_type::max_size;
-    using allocator_type::address;
-    using allocator_type::construct;
-    using allocator_type::destroy;
-#else
+
+    typedef typename std::add_lvalue_reference<value_type>::type &reference;
+    typedef typename std::add_lvalue_reference<typename std::add_const<value_type>::type>::type &const_reference;
+
+    pointer address(reference r) const RAPIDJSON_NOEXCEPT
+    {
+        return std::addressof(r);
+    }
+    const_pointer address(const_reference r) const RAPIDJSON_NOEXCEPT
+    {
+        return std::addressof(r);
+    }
+
     size_t max_size() const RAPIDJSON_NOEXCEPT
     {
-        return allocator_type::max_size();
+        return std::numeric_limits<size_type>::max() / sizeof(value_type);
```

Comment:
Why not do `return traits_type::max_size(*this);`? Same for `construct` and `destroy`.

Reply:
The documentation [1] seems to suggest that std::allocator_traits::max_size() calls the passed in allocator max_size() member if available, that'd be an infinite recursion I suppose here..

[1] https://en.cppreference.com/w/cpp/memory/allocator_traits/max_size

Reply:
We'd be using the allocator_traits for std::allocator<T>, not StdAllocator<T>, so it should work fine.

Reply:
Hm ok, `traits_type::max_size(*this)` would be using `(*this).std::allocator::max_size()`.
That's tricky though, and since std::allocator::max_size() is quite unlikely to return something else than the above I'd rather avoid it.

Given that my code does `typedef allocator_type traits_type` for C++ < 11, we could
```
    size_t max_size() const RAPIDJSON_NOEXCEPT
    {
        return traits_type::max_size();
    }
```
regardless of C++ version, but is it worth it? Some special MSVC's std::allocator::max_size() again?

Reply:
That wouldn't work because allocator_traits wants us to pass in an instance of the allocator while calling the base class's member doesn't.

I think the existing code is fine, just a nit.

Reply:
Ah yes, right. Thanks.

Reply:
Finally I used the traits_type version of `max_size`, `construct` and `destroy` because MSVC somehow didn't like my explicit `p->~T()` in `destroy` (see [1]). So let it use its own implementation..

[1] https://ci.appveyor.com/project/miloyip/rapidjson-0fdqj/builds/38443196/job/m66fsm9nbyte8vvx#L247